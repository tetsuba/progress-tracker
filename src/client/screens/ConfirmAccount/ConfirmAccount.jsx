import React, { useEffect, useState } from 'react'
import { Link, useParams } from 'react-router-dom'
import { useQuery } from '@apollo/react-hooks'

// COMPONENTS
import { Col, Container, Row } from 'react-bootstrap'
import EmailVerificationForm from '../../components/Form/EmailVerificationForm'
import Loading from '../../components/Loading/Loading'

// QUERY
import { VALIDATE_USER_EMAIL_QUERY } from '../../api/user/user.query'

// TYPES
type Props = {
  pageState: string,
}

export default function ConfirmAccount(props: Props) {
  const { token } = useParams()
  const variables = { token: token ? decodeURIComponent(token) : '' }
  const { error, data } = useQuery(VALIDATE_USER_EMAIL_QUERY, { variables })
  const [pageState, setPageSate] = useState(props.pageState)

  useEffect(() => {
    if (data) setPageSate('accountVerified')
    if (error) setPageSate('tokenExpired')
  }, [error, data])

  return (
    <Container className="pt-5">
      <Row>
        {
          {
            loading: <Loading />,
            tokenExpired: (
              <EmailVerificationForm
                setPageState={(state) => setPageSate(state)}
              >
                <h3 className="text-danger">
                  Email verification session expired
                </h3>
                <p>
                  Please enter your email address below and a link will be sent
                  to confirm your email address.
                </p>
              </EmailVerificationForm>
            ),
            accountVerified: (
              <Col className="text-center">
                <h3>Email Verified</h3>
                <p>Please log into your account</p>
                <Link to="/login">Login</Link>
              </Col>
            ),
            success: (
              <Col id="LoginSuccess">
                <h3>Please check your email.</h3>
              </Col>
            ),
          }[pageState]
        }
      </Row>
    </Container>
  )
}

ConfirmAccount.defaultProps = {
  pageState: 'loading',
}
