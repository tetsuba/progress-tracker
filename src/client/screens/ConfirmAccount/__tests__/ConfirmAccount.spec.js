import { act } from 'react-dom/test-utils'
import { delay, graphRenderer } from '../../../../test/testHelper'
import ConfirmAccount from '../ConfirmAccount'

import {
  validateUserEmailQuerySuccess,
  validateUserEmailQueryError,
} from '../../../../test/mockApi/user/userMockQuery'

jest.mock('react-router-dom', () => ({
  useParams: jest.fn().mockReturnValue({ token: 'confirmToken1234' }),
  Link: ({ children }) => children,
}))

describe('<ConfirmAccount>', () => {
  describe('@Render', () => {
    it('should render "loading"', async () => {
      let wrapper
      await act(async () => {
        wrapper = graphRenderer(
          ConfirmAccount,
          [validateUserEmailQuerySuccess],
          { pageState: 'loading' }
        )
        await delay()
      })
      expect(wrapper.find(ConfirmAccount)).toMatchSnapshot()
    })
    it('should render "tokenExpired"', async () => {
      let wrapper
      await act(async () => {
        wrapper = graphRenderer(
          ConfirmAccount,
          [validateUserEmailQuerySuccess],
          { pageState: 'tokenExpired' }
        )
        await delay()
      })
      expect(wrapper.find(ConfirmAccount)).toMatchSnapshot()
    })
    it('should render "accountVerified"', async () => {
      let wrapper
      await act(async () => {
        wrapper = graphRenderer(
          ConfirmAccount,
          [validateUserEmailQuerySuccess],
          { pageState: 'accountVerified' }
        )
        await delay()
      })
      expect(wrapper.find(ConfirmAccount)).toMatchSnapshot()
    })
    it('should render "success"', async () => {
      let wrapper
      await act(async () => {
        wrapper = graphRenderer(
          ConfirmAccount,
          [validateUserEmailQuerySuccess],
          { pageState: 'success' }
        )
        await delay()
      })
      expect(wrapper.find(ConfirmAccount)).toMatchSnapshot()
    })
  })

  describe('@Query', () => {
    describe('Success', () => {
      it('should render "accountVerified"', async () => {
        let wrapper
        await act(async () => {
          wrapper = graphRenderer(
            ConfirmAccount,
            [validateUserEmailQuerySuccess],
            {}
          )
          await delay()
        })
        wrapper.update()
        expect(wrapper.find(ConfirmAccount)).toMatchSnapshot()
      })
    })
    describe('Error', () => {
      it('should render "tokenExpired"', async () => {
        let wrapper
        await act(async () => {
          wrapper = graphRenderer(
            ConfirmAccount,
            [validateUserEmailQueryError],
            {}
          )
          await delay()
        })
        wrapper.update()
        expect(wrapper.find(ConfirmAccount)).toMatchSnapshot()
      })
    })
  })
})
