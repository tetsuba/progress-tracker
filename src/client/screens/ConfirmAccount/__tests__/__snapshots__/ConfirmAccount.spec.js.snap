// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfirmAccount> @Query Error should render "tokenExpired" 1`] = `
<ConfirmAccount
  pageState="loading"
>
  <Container
    className="pt-5"
    fluid={false}
  >
    <div
      className="pt-5 container"
    >
      <Row
        noGutters={false}
      >
        <div
          className="row"
        >
          <EmailVerificationForm
            setPageState={[Function]}
          >
            <Box
              max={500}
            >
              <div
                style={
                  Object {
                    "backgroundColor": "#f8f9fa",
                    "border": "solid 1px #ccc",
                    "borderRadius": "8px",
                    "margin": "0 auto",
                    "maxWidth": 500,
                    "padding": "20px 20px",
                    "width": "100%",
                  }
                }
              >
                <Col>
                  <div
                    className="col"
                  >
                    <h3
                      className="text-danger"
                    >
                      Email verification session expired
                    </h3>
                    <p>
                      Please enter your email address below and a link will be sent to confirm your email address.
                    </p>
                    <Form
                      id="EmailVerificationForm"
                      inline={false}
                      onSubmit={[Function]}
                    >
                      <form
                        className=""
                        id="EmailVerificationForm"
                        onSubmit={[Function]}
                      >
                        <FormGroup
                          controlId="EmailVerificationEmail"
                        >
                          <div
                            className="form-group"
                          >
                            <FormControl
                              isInvalid={false}
                              name="email"
                              onChange={[Function]}
                              placeholder="email@address.com"
                              required={true}
                              type="email"
                              value=""
                            >
                              <input
                                className="form-control"
                                id="EmailVerificationEmail"
                                name="email"
                                onChange={[Function]}
                                placeholder="email@address.com"
                                required={true}
                                type="email"
                                value=""
                              />
                            </FormControl>
                            <Feedback
                              type="invalid"
                            >
                              <div
                                className="invalid-feedback"
                              />
                            </Feedback>
                            <FormText
                              className="text-muted"
                            >
                              <small
                                className="text-muted form-text"
                              >
                                We'll never share your email with anyone else.
                              </small>
                            </FormText>
                          </div>
                        </FormGroup>
                        <Button
                          active={false}
                          className="float-right"
                          disabled={false}
                          id="EmailVerificationSubmit"
                          type="submit"
                          variant="primary"
                        >
                          <button
                            className="float-right btn btn-primary"
                            disabled={false}
                            id="EmailVerificationSubmit"
                            type="submit"
                          >
                            Send
                          </button>
                        </Button>
                      </form>
                    </Form>
                    <Link
                      to="/"
                    >
                      <FontAwesomeIcon
                        border={false}
                        className=""
                        fixedWidth={false}
                        flip={null}
                        icon={
                          Object {
                            "icon": Array [
                              192,
                              512,
                              Array [],
                              "f0d9",
                              "M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z",
                            ],
                            "iconName": "caret-left",
                            "prefix": "fas",
                          }
                        }
                        inverse={false}
                        listItem={false}
                        mask={null}
                        pull={null}
                        pulse={false}
                        rotation={null}
                        size={null}
                        spin={false}
                        swapOpacity={false}
                        symbol={false}
                        title=""
                        transform={null}
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-caret-left fa-w-6 "
                          data-icon="caret-left"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          style={Object {}}
                          viewBox="0 0 192 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </FontAwesomeIcon>
                       back
                    </Link>
                  </div>
                </Col>
              </div>
            </Box>
          </EmailVerificationForm>
        </div>
      </Row>
    </div>
  </Container>
</ConfirmAccount>
`;

exports[`<ConfirmAccount> @Query Success should render "accountVerified" 1`] = `
<ConfirmAccount
  pageState="loading"
>
  <Container
    className="pt-5"
    fluid={false}
  >
    <div
      className="pt-5 container"
    >
      <Row
        noGutters={false}
      >
        <div
          className="row"
        >
          <Col
            className="text-center"
          >
            <div
              className="text-center col"
            >
              <h3>
                Email Verified
              </h3>
              <p>
                Please log into your account
              </p>
              <Link
                to="/login"
              >
                Login
              </Link>
            </div>
          </Col>
        </div>
      </Row>
    </div>
  </Container>
</ConfirmAccount>
`;

exports[`<ConfirmAccount> @Render should render "accountVerified" 1`] = `
<ConfirmAccount
  pageState="accountVerified"
>
  <Container
    className="pt-5"
    fluid={false}
  >
    <div
      className="pt-5 container"
    >
      <Row
        noGutters={false}
      >
        <div
          className="row"
        >
          <Col
            className="text-center"
          >
            <div
              className="text-center col"
            >
              <h3>
                Email Verified
              </h3>
              <p>
                Please log into your account
              </p>
              <Link
                to="/login"
              >
                Login
              </Link>
            </div>
          </Col>
        </div>
      </Row>
    </div>
  </Container>
</ConfirmAccount>
`;

exports[`<ConfirmAccount> @Render should render "loading" 1`] = `
<ConfirmAccount
  pageState="loading"
>
  <Container
    className="pt-5"
    fluid={false}
  >
    <div
      className="pt-5 container"
    >
      <Row
        noGutters={false}
      >
        <div
          className="row"
        >
          <Loading>
            <Fade
              appear={false}
              in={false}
              mountOnEnter={false}
              timeout={300}
              unmountOnExit={false}
            >
              <Transition
                addEndListener={[Function]}
                appear={false}
                enter={true}
                exit={true}
                in={false}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={300}
                unmountOnExit={false}
              >
                <div
                  className="fade"
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "rgba(0,0,0,0.5)",
                      "display": "flex",
                      "height": "100%",
                      "justifyContent": "center",
                      "left": "0",
                      "position": "absolute",
                      "top": "0",
                      "width": "100%",
                      "zIndex": "1000",
                    }
                  }
                >
                  <Spinner
                    animation="border"
                    role="status"
                    variant="light"
                  >
                    <div
                      className="spinner-border text-light"
                      role="status"
                    >
                      <span
                        className="sr-only"
                      >
                        Loading...
                      </span>
                    </div>
                  </Spinner>
                </div>
              </Transition>
            </Fade>
          </Loading>
        </div>
      </Row>
    </div>
  </Container>
</ConfirmAccount>
`;

exports[`<ConfirmAccount> @Render should render "success" 1`] = `
<ConfirmAccount
  pageState="success"
>
  <Container
    className="pt-5"
    fluid={false}
  >
    <div
      className="pt-5 container"
    >
      <Row
        noGutters={false}
      >
        <div
          className="row"
        >
          <Col
            id="LoginSuccess"
          >
            <div
              className="col"
              id="LoginSuccess"
            >
              <h3>
                Please check your email.
              </h3>
            </div>
          </Col>
        </div>
      </Row>
    </div>
  </Container>
</ConfirmAccount>
`;

exports[`<ConfirmAccount> @Render should render "tokenExpired" 1`] = `
<ConfirmAccount
  pageState="tokenExpired"
>
  <Container
    className="pt-5"
    fluid={false}
  >
    <div
      className="pt-5 container"
    >
      <Row
        noGutters={false}
      >
        <div
          className="row"
        >
          <EmailVerificationForm
            setPageState={[Function]}
          >
            <Box
              max={500}
            >
              <div
                style={
                  Object {
                    "backgroundColor": "#f8f9fa",
                    "border": "solid 1px #ccc",
                    "borderRadius": "8px",
                    "margin": "0 auto",
                    "maxWidth": 500,
                    "padding": "20px 20px",
                    "width": "100%",
                  }
                }
              >
                <Col>
                  <div
                    className="col"
                  >
                    <h3
                      className="text-danger"
                    >
                      Email verification session expired
                    </h3>
                    <p>
                      Please enter your email address below and a link will be sent to confirm your email address.
                    </p>
                    <Form
                      id="EmailVerificationForm"
                      inline={false}
                      onSubmit={[Function]}
                    >
                      <form
                        className=""
                        id="EmailVerificationForm"
                        onSubmit={[Function]}
                      >
                        <FormGroup
                          controlId="EmailVerificationEmail"
                        >
                          <div
                            className="form-group"
                          >
                            <FormControl
                              isInvalid={false}
                              name="email"
                              onChange={[Function]}
                              placeholder="email@address.com"
                              required={true}
                              type="email"
                              value=""
                            >
                              <input
                                className="form-control"
                                id="EmailVerificationEmail"
                                name="email"
                                onChange={[Function]}
                                placeholder="email@address.com"
                                required={true}
                                type="email"
                                value=""
                              />
                            </FormControl>
                            <Feedback
                              type="invalid"
                            >
                              <div
                                className="invalid-feedback"
                              />
                            </Feedback>
                            <FormText
                              className="text-muted"
                            >
                              <small
                                className="text-muted form-text"
                              >
                                We'll never share your email with anyone else.
                              </small>
                            </FormText>
                          </div>
                        </FormGroup>
                        <Button
                          active={false}
                          className="float-right"
                          disabled={false}
                          id="EmailVerificationSubmit"
                          type="submit"
                          variant="primary"
                        >
                          <button
                            className="float-right btn btn-primary"
                            disabled={false}
                            id="EmailVerificationSubmit"
                            type="submit"
                          >
                            Send
                          </button>
                        </Button>
                      </form>
                    </Form>
                    <Link
                      to="/"
                    >
                      <FontAwesomeIcon
                        border={false}
                        className=""
                        fixedWidth={false}
                        flip={null}
                        icon={
                          Object {
                            "icon": Array [
                              192,
                              512,
                              Array [],
                              "f0d9",
                              "M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z",
                            ],
                            "iconName": "caret-left",
                            "prefix": "fas",
                          }
                        }
                        inverse={false}
                        listItem={false}
                        mask={null}
                        pull={null}
                        pulse={false}
                        rotation={null}
                        size={null}
                        spin={false}
                        swapOpacity={false}
                        symbol={false}
                        title=""
                        transform={null}
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-caret-left fa-w-6 "
                          data-icon="caret-left"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          style={Object {}}
                          viewBox="0 0 192 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </FontAwesomeIcon>
                       back
                    </Link>
                  </div>
                </Col>
              </div>
            </Box>
          </EmailVerificationForm>
        </div>
      </Row>
    </div>
  </Container>
</ConfirmAccount>
`;
